File: modules/omnitak_mobile/src/valdi/omnitak/services/MultiServerFederation.ts
/**
 * Multi-Server Federation Manager
 *
 * Manages multiple simultaneous TAK server connections with:
 * - Federated data collection from all servers
 * - Selective data sharing to blue team (friendly forces)
 * - Deduplication of CoT events across servers
 * - Per-server data sharing policies
 */

import { takService, ServerConfig, ConnectionInfo } from './TakService';
import { parseCotXml, CotEvent } from './CotParser';

/**
 * Data types that can be selectively shared
 */
export type DataType =
  | 'friendly'      // a-f-* - Friendly forces
  | 'hostile'       // a-h-* - Hostile forces
  | 'unknown'       // a-u-* - Unknown forces
  | 'neutral'       // a-n-* - Neutral forces
  | 'sensor'        // b-* - Sensor data
  | 'geofence'      // u-d-f - Geofence/boundaries
  | 'route'         // b-m-p-c - Route planning
  | 'casevac'       // b-r-f-h-c - CASEVAC requests
  | 'target'        // u-d-c-c - Target designation
  | 'all';          // All data types

/**
 * @ExportModel({
 *   ios: 'DataSharingPolicy',
 *   android: 'com.engindearing.omnitak.DataSharingPolicy'
 * })
 */
export interface DataSharingPolicy {
  // Which data types to receive from this server
  receiveTypes: string[];

  // Which data types to send to this server
  sendTypes: string[];

  // Auto-share mode: automatically share received data to other servers
  autoShare: boolean;

  // Blue team mode: only share friendly data
  blueTeamOnly: boolean;

  // Bidirectional: both send and receive (typical)
  bidirectional: boolean;
}

/**
 * @ExportModel({
 *   ios: 'FederatedServer',
 *   android: 'com.engindearing.omnitak.FederatedServer'
 * })
 */
export interface FederatedServer {
  id: string;
  name: string;
  connectionId: number | null;
  config: ServerConfig;
  policy: DataSharingPolicy;
  status: string;
  lastError?: string;
}

/**
 * @ExportModel({
 *   ios: 'FederatedCoTEvent',
 *   android: 'com.engindearing.omnitak.FederatedCoTEvent'
 * })
 */
export interface FederatedCoTEvent {
  event: CotEvent;
  sourceServerId: string;
  sourceServerName: string;
  receivedAt: number; // Unix timestamp in milliseconds
  sharedTo: string[]; // Server IDs this event has been shared to
}

/**
 * Multi-Server Federation Manager
 */
export class MultiServerFederation {
  private servers: Map<string, FederatedServer> = new Map();
  private eventCache: Map<string, FederatedCoTEvent> = new Map(); // UID -> Event
  private eventCallbacks: Set<(event: FederatedCoTEvent) => void> = new Set();
  private statusCallbacks: Set<() => void> = new Set();
  private cotUnsubscribers: Map<number, () => void> = new Map();

  constructor() {
    console.log('MultiServerFederation initialized');
  }

  /**
   * Add a server to the federation
   */
  addServer(
    id: string,
    name: string,
    config: ServerConfig,
    policy?: Partial<DataSharingPolicy>
  ): void {
    const defaultPolicy: DataSharingPolicy = {
      receiveTypes: ['all'],
      sendTypes: ['friendly'], // Default: only send friendly data
      autoShare: true,
      blueTeamOnly: true, // Default: blue team mode on
      bidirectional: true,
    };

    const server: FederatedServer = {
      id,
      name,
      connectionId: null,
      config,
      policy: { ...defaultPolicy, ...policy },
      status: 'disconnected',
    };

    this.servers.set(id, server);
    console.log(`Added server to federation: ${name} (${id})`);
    this.notifyStatusChange();
  }

  /**
   * Remove a server from the federation
   */
  async removeServer(id: string): Promise<void> {
    const server = this.servers.get(id);
    if (server && server.connectionId !== null) {
      await this.disconnectServer(id);
    }
    this.servers.delete(id);
    console.log(`Removed server from federation: ${id}`);
    this.notifyStatusChange();
  }

  /**
   * Update server sharing policy
   */
  updatePolicy(id: string, policy: Partial<DataSharingPolicy>): void {
    const server = this.servers.get(id);
    if (server) {
      server.policy = { ...server.policy, ...policy };
      console.log(`Updated policy for server: ${id}`, policy);
      this.notifyStatusChange();
    }
  }

  /**
   * Connect to a server
   */
  async connectServer(id: string): Promise<boolean> {
    const server = this.servers.get(id);
    if (!server) {
      console.error(`Server not found: ${id}`);
      return false;
    }

    if (server.connectionId !== null) {
      console.warn(`Server already connected: ${id}`);
      return true;
    }

    server.status = 'connecting';
    this.notifyStatusChange();

    try {
      const connectionId = await takService.connect(server.config);
      if (connectionId === null) {
        server.status = 'error';
        server.lastError = 'Failed to connect';
        this.notifyStatusChange();
        return false;
      }

      server.connectionId = connectionId;
      server.status = 'connected';

      // Subscribe to CoT messages from this server
      const unsubscribe = takService.onCotReceived(connectionId, (xml: string) => {
        this.handleIncomingCot(id, xml);
      });
      this.cotUnsubscribers.set(connectionId, unsubscribe);

      console.log(`Connected to server: ${server.name} (${id})`);
      this.notifyStatusChange();
      return true;
    } catch (error) {
      server.status = 'error';
      server.lastError = String(error);
      console.error(`Error connecting to server ${id}:`, error);
      this.notifyStatusChange();
      return false;
    }
  }

  /**
   * Disconnect from a server
   */
  async disconnectServer(id: string): Promise<void> {
    const server = this.servers.get(id);
    if (!server || server.connectionId === null) {
      return;
    }

    const connectionId = server.connectionId;

    // Unsubscribe from CoT messages
    const unsubscribe = this.cotUnsubscribers.get(connectionId);
    if (unsubscribe) {
      unsubscribe();
      this.cotUnsubscribers.delete(connectionId);
    }

    await takService.disconnect(connectionId);
    server.connectionId = null;
    server.status = 'disconnected';

    console.log(`Disconnected from server: ${server.name} (${id})`);
    this.notifyStatusChange();
  }

  /**
   * Connect to all servers in the federation
   */
  async connectAll(): Promise<void> {
    const promises = Array.from(this.servers.keys()).map((id) =>
      this.connectServer(id)
    );
    await Promise.all(promises);
  }

  /**
   * Disconnect from all servers
   */
  async disconnectAll(): Promise<void> {
    const promises = Array.from(this.servers.keys()).map((id) =>
      this.disconnectServer(id)
    );
    await Promise.all(promises);
  }

  /**
   * Handle incoming CoT message from a server
   */
  private handleIncomingCot(serverId: string, xml: string): void {
    const server = this.servers.get(serverId);
    if (!server) return;

    const event = parseCotXml(xml);
    if (!event) return;

    // Check if this data type should be received from this server
    if (!this.shouldReceive(server, event)) {
      console.log(`Filtered incoming event from ${server.name}: ${event.type}`);
      return;
    }

    // Create or update federated event
    const existingEvent = this.eventCache.get(event.uid);
    const federatedEvent: FederatedCoTEvent = existingEvent || {
      event,
      sourceServerId: serverId,
      sourceServerName: server.name,
      receivedAt: Date.now(),
      sharedTo: [],
    };

    // Update event data if newer
    federatedEvent.event = event;

    this.eventCache.set(event.uid, federatedEvent);

    // Notify subscribers
    this.eventCallbacks.forEach((callback) => {
      try {
        callback(federatedEvent);
      } catch (error) {
        console.error('Error in event callback:', error);
      }
    });

    // Auto-share to other servers if policy allows
    if (server.policy.autoShare) {
      this.shareEventToOtherServers(federatedEvent, serverId);
    }
  }

  /**
   * Check if data should be received from this server based on policy
   */
  private shouldReceive(server: FederatedServer, event: CotEvent): boolean {
    const { receiveTypes } = server.policy;

    if (receiveTypes.includes('all')) {
      return true;
    }

    const dataType = this.getDataType(event.type);
    return receiveTypes.includes(dataType);
  }

  /**
   * Check if data should be sent to this server based on policy
   */
  private shouldSend(server: FederatedServer, event: CotEvent): boolean {
    const { sendTypes, blueTeamOnly } = server.policy;

    // Blue team only mode: only send friendly data
    if (blueTeamOnly && !event.type.includes('a-f-')) {
      return false;
    }

    if (sendTypes.includes('all')) {
      return true;
    }

    const dataType = this.getDataType(event.type);
    return sendTypes.includes(dataType);
  }

  /**
   * Share event to other servers based on their policies
   */
  private async shareEventToOtherServers(
    federatedEvent: FederatedCoTEvent,
    sourceServerId: string
  ): Promise<void> {
    const { event, sharedTo } = federatedEvent;

    for (const [id, server] of this.servers) {
      // Skip source server and already shared servers
      if (id === sourceServerId || sharedTo.includes(id)) {
        continue;
      }

      // Skip if not connected
      if (server.status !== 'connected' || server.connectionId === null) {
        continue;
      }

      // Check if this server should receive this data
      if (!this.shouldSend(server, event)) {
        console.log(`Not sharing event to ${server.name}: policy restriction`);
        continue;
      }

      // Generate CoT XML and send
      const cotXml = this.generateCotXml(event);
      const success = await takService.sendCot(server.connectionId, cotXml);

      if (success) {
        sharedTo.push(id);
        console.log(`Shared event ${event.uid} to ${server.name}`);
      }
    }
  }

  /**
   * Manually send CoT event to specific servers
   */
  async sendToServers(event: CotEvent, serverIds: string[]): Promise<void> {
    const cotXml = this.generateCotXml(event);

    for (const id of serverIds) {
      const server = this.servers.get(id);
      if (!server || server.connectionId === null) {
        console.warn(`Cannot send to server ${id}: not connected`);
        continue;
      }

      if (!this.shouldSend(server, event)) {
        console.warn(`Cannot send to server ${id}: policy restriction`);
        continue;
      }

      await takService.sendCot(server.connectionId, cotXml);
    }
  }

  /**
   * Broadcast CoT event to all connected servers (respecting policies)
   */
  async broadcast(event: CotEvent): Promise<void> {
    const serverIds = Array.from(this.servers.keys());
    await this.sendToServers(event, serverIds);
  }

  /**
   * Get federated data type from CoT type string
   */
  private getDataType(cotType: string): DataType {
    if (cotType.startsWith('a-f-')) return 'friendly';
    if (cotType.startsWith('a-h-')) return 'hostile';
    if (cotType.startsWith('a-u-')) return 'unknown';
    if (cotType.startsWith('a-n-')) return 'neutral';
    if (cotType.startsWith('b-m-p-c')) return 'route';
    if (cotType.startsWith('b-r-f-h-c')) return 'casevac';
    if (cotType.startsWith('u-d-f')) return 'geofence';
    if (cotType.startsWith('u-d-c-c')) return 'target';
    if (cotType.startsWith('b-')) return 'sensor';
    return 'unknown';
  }

  /**
   * Generate CoT XML from event
   */
  private generateCotXml(event: CotEvent): string {
    // Generate CoT XML - simplified version
    const callsign = event.detail?.contact?.callsign || event.uid;
    const team = event.detail?.group?.name || 'Cyan';

    return `<?xml version="1.0" encoding="UTF-8"?>
<event version="2.0" uid="${event.uid}" type="${event.type}" time="${event.time}" start="${event.start}" stale="${event.stale}" how="h-e">
  <point lat="${event.point.lat}" lon="${event.point.lon}" hae="${event.point.hae}" ce="9999999" le="9999999"/>
  <detail>
    <contact callsign="${callsign}"/>
    <__group name="${team}" role="Team Member"/>
  </detail>
</event>`;
  }

  /**
   * Subscribe to federated events
   */
  onFederatedEvent(callback: (event: FederatedCoTEvent) => void): () => void {
    this.eventCallbacks.add(callback);
    return () => this.eventCallbacks.delete(callback);
  }

  /**
   * Subscribe to status changes
   */
  onStatusChange(callback: () => void): () => void {
    this.statusCallbacks.add(callback);
    return () => this.statusCallbacks.delete(callback);
  }

  /**
   * Get all federated events
   */
  getFederatedEvents(): FederatedCoTEvent[] {
    return Array.from(this.eventCache.values());
  }

  /**
   * Get all servers
   */
  getServers(): FederatedServer[] {
    return Array.from(this.servers.values());
  }

  /**
   * Get server by ID
   */
  getServer(id: string): FederatedServer | undefined {
    return this.servers.get(id);
  }

  /**
   * Get connected servers count
   */
  getConnectedCount(): number {
    return Array.from(this.servers.values()).filter(
      (s) => s.status === 'connected'
    ).length;
  }

  /**
   * Clear event cache
   */
  clearCache(): void {
    this.eventCache.clear();
    console.log('Event cache cleared');
  }

  /**
   * Notify status change subscribers
   */
  private notifyStatusChange(): void {
    this.statusCallbacks.forEach((callback) => {
      try {
        callback();
      } catch (error) {
        console.error('Error in status callback:', error);
      }
    });
  }
}

// Singleton instance
export const multiServerFederation = new MultiServerFederation();

File: apps/omnitak/OmniTAKMobile/Utilities/Network/MultiServerFederation.swift
import Foundation
import Combine
import CoreLocation

// MARK: - Data Types

enum DataType: String, CaseIterable {
    case friendly = "friendly"      // a-f-*
    case hostile = "hostile"        // a-h-*
    case unknown = "unknown"        // a-u-*
    case neutral = "neutral"        // a-n-*
    case sensor = "sensor"          // b-*
    case geofence = "geofence"      // u-d-f
    case route = "route"            // b-m-p-c
    case casevac = "casevac"        // b-r-f-h-c
    case target = "target"          // u-d-c-c
    case all = "all"
}

// MARK: - Data Sharing Policy

struct DataSharingPolicy {
    var receiveTypes: [DataType]
    var sendTypes: [DataType]
    var autoShare: Bool
    var blueTeamOnly: Bool
    var bidirectional: Bool

    static var `default`: DataSharingPolicy {
        DataSharingPolicy(
            receiveTypes: [.all],
            sendTypes: [.friendly],  // Default: only send friendly data
            autoShare: true,
            blueTeamOnly: true,      // Default: blue team mode on
            bidirectional: true
        )
    }
}

// MARK: - Federated Server

struct FederatedServer: Identifiable {
    let id: String
    var name: String
    var host: String
    var port: UInt16
    var protocolType: String
    var useTLS: Bool
    var certificateName: String?
    var certificatePassword: String?
    var policy: DataSharingPolicy
    var status: ServerStatus
    var lastError: String?
    var takService: TAKService?

    enum ServerStatus: String {
        case connected
        case connecting
        case disconnected
        case error
    }
}

// MARK: - Federated CoT Event

struct FederatedCoTEvent {
    let event: CoTEvent
    let sourceServerId: String
    let sourceServerName: String
    let receivedAt: Date
    var sharedTo: [String]  // Server IDs this event has been shared to
}

// MARK: - Multi-Server Federation Manager

class MultiServerFederation: ObservableObject {
    @Published var servers: [FederatedServer] = []
    @Published var federatedEvents: [String: FederatedCoTEvent] = [:]  // UID -> Event

    private var cancellables = Set<AnyCancellable>()

    init() {
        #if DEBUG
        print("🌐 MultiServerFederation initialized")
        #endif
    }

    // MARK: - Server Management

    func addServer(
        id: String,
        name: String,
        host: String,
        port: UInt16,
        protocolType: String,
        useTLS: Bool,
        certificateName: String? = nil,
        certificatePassword: String? = nil,
        policy: DataSharingPolicy = .default
    ) {
        let server = FederatedServer(
            id: id,
            name: name,
            host: host,
            port: port,
            protocolType: protocolType,
            useTLS: useTLS,
            certificateName: certificateName,
            certificatePassword: certificatePassword,
            policy: policy,
            status: .disconnected,
            takService: nil
        )

        servers.append(server)
        #if DEBUG
        print("✅ Added server to federation: \(name) (\(id))")
        #endif
    }

    func removeServer(id: String) {
        if let index = servers.firstIndex(where: { $0.id == id }) {
            let server = servers[index]
            if server.status == .connected {
                disconnectServer(id: id)
            }
            servers.remove(at: index)
            #if DEBUG
            print("🗑️ Removed server from federation: \(id)")
            #endif
        }
    }

    func updatePolicy(id: String, policy: DataSharingPolicy) {
        if let index = servers.firstIndex(where: { $0.id == id }) {
            servers[index].policy = policy
            print("📋 Updated policy for server: \(id)")
        }
    }

    // MARK: - Connection Management

    func connectServer(id: String) {
        guard let index = servers.firstIndex(where: { $0.id == id }) else {
            print("❌ Server not found: \(id)")
            return
        }

        let server = servers[index]
        if server.status == .connected {
            #if DEBUG
            print("⚠️ Server already connected: \(id)")
            #endif
            return
        }

        servers[index].status = .connecting

        // Create TAKService instance
        let takService = TAKService()
        servers[index].takService = takService

        // Subscribe to CoT events from this server
        takService.onCoTReceived = { [weak self] event in
            self?.handleIncomingCoT(serverId: id, event: event)
        }

        // Connect to server with certificate support
        takService.connect(
            host: server.host,
            port: server.port,
            protocolType: server.protocolType,
            useTLS: server.useTLS,
            certificateName: server.certificateName,
            certificatePassword: server.certificatePassword
        )

        // Update status based on connection result
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { [weak self] in
            guard let self = self else { return }
            if takService.isConnected {
                if let idx = self.servers.firstIndex(where: { $0.id == id }) {
                    self.servers[idx].status = .connected
                    #if DEBUG
                    print("✅ Connected to server: \(server.name) (\(id))")
                    #endif
                }
            } else {
                if let idx = self.servers.firstIndex(where: { $0.id == id }) {
                    self.servers[idx].status = .error
                    self.servers[idx].lastError = "Failed to connect"
                    print("❌ Failed to connect to server: \(id)")
                }
            }
        }
    }

    func disconnectServer(id: String) {
        guard let index = servers.firstIndex(where: { $0.id == id }) else {
            return
        }

        servers[index].takService?.disconnect()
        servers[index].takService = nil
        servers[index].status = .disconnected

        #if DEBUG
        print("🔌 Disconnected from server: \(id)")
        #endif
    }

    func connectAll() {
        print("⚡ Connecting to all servers...")
        for server in servers {
            if server.status == .disconnected {
                connectServer(id: server.id)
            }
        }
    }

    func disconnectAll() {
        print("⏸ Disconnecting from all servers...")
        for server in servers {
            if server.status == .connected {
                disconnectServer(id: server.id)
            }
        }
    }

    // MARK: - Data Federation

    private func handleIncomingCoT(serverId: String, event: CoTEvent) {
        guard let server = servers.first(where: { $0.id == serverId }) else {
            return
        }

        // Check if this data type should be received from this server
        if !shouldReceive(server: server, event: event) {
            print("🚫 Filtered incoming event from \(server.name): \(event.type)")
            return
        }

        // Create or update federated event
        let federatedEvent: FederatedCoTEvent
        if let existing = federatedEvents[event.uid] {
            federatedEvent = FederatedCoTEvent(
                event: event,
                sourceServerId: existing.sourceServerId,
                sourceServerName: existing.sourceServerName,
                receivedAt: existing.receivedAt,
                sharedTo: existing.sharedTo
            )
        } else {
            federatedEvent = FederatedCoTEvent(
                event: event,
                sourceServerId: serverId,
                sourceServerName: server.name,
                receivedAt: Date(),
                sharedTo: []
            )
        }

        federatedEvents[event.uid] = federatedEvent

        // Auto-share to other servers if policy allows
        if server.policy.autoShare {
            shareEventToOtherServers(event: federatedEvent, sourceServerId: serverId)
        }
    }

    private func shouldReceive(server: FederatedServer, event: CoTEvent) -> Bool {
        let policy = server.policy

        if policy.receiveTypes.contains(.all) {
            return true
        }

        let dataType = getDataType(cotType: event.type)
        return policy.receiveTypes.contains(dataType)
    }

    private func shouldSend(server: FederatedServer, event: CoTEvent) -> Bool {
        let policy = server.policy

        // Blue team only mode: only send friendly data
        if policy.blueTeamOnly && !event.type.contains("a-f-") {
            return false
        }

        if policy.sendTypes.contains(.all) {
            return true
        }

        let dataType = getDataType(cotType: event.type)
        return policy.sendTypes.contains(dataType)
    }

    private func shareEventToOtherServers(event: FederatedCoTEvent, sourceServerId: String) {
        let cotXml = generateCoTXml(event: event.event)

        for server in servers {
            // Skip source server and already shared servers
            if server.id == sourceServerId || event.sharedTo.contains(server.id) {
                continue
            }

            // Skip if not connected
            if server.status != .connected || server.takService == nil {
                continue
            }

            // Check if this server should receive this data
            if !shouldSend(server: server, event: event.event) {
                print("🚫 Not sharing event to \(server.name): policy restriction")
                continue
            }

            // Send to server
            if let takService = server.takService {
                let success = takService.sendCoT(xml: cotXml)
                if success {
                    federatedEvents[event.event.uid]?.sharedTo.append(server.id)
                    #if DEBUG
                    print("✅ Shared event \(event.event.uid) to \(server.name)")
                    #endif
                }
            }
        }
    }

    // MARK: - Manual Sending

    func sendToServers(event: CoTEvent, serverIds: [String]) {
        let cotXml = generateCoTXml(event: event)

        for serverId in serverIds {
            guard let server = servers.first(where: { $0.id == serverId }) else {
                #if DEBUG
                print("⚠️ Cannot send to server \(serverId): not found")
                #endif
                continue
            }

            guard server.status == .connected, let takService = server.takService else {
                #if DEBUG
                print("⚠️ Cannot send to server \(serverId): not connected")
                #endif
                continue
            }

            if !shouldSend(server: server, event: event) {
                #if DEBUG
                print("⚠️ Cannot send to server \(serverId): policy restriction")
                #endif
                continue
            }

            _ = takService.sendCoT(xml: cotXml)
        }
    }

    func broadcast(event: CoTEvent) {
        let serverIds = servers.map { $0.id }
        sendToServers(event: event, serverIds: serverIds)
    }

    // MARK: - Utility Functions

    private func getDataType(cotType: String) -> DataType {
        if cotType.hasPrefix("a-f-") { return .friendly }
        if cotType.hasPrefix("a-h-") { return .hostile }
        if cotType.hasPrefix("a-u-") { return .unknown }
        if cotType.hasPrefix("a-n-") { return .neutral }
        if cotType.hasPrefix("b-m-p-c") { return .route }
        if cotType.hasPrefix("b-r-f-h-c") { return .casevac }
        if cotType.hasPrefix("u-d-f") { return .geofence }
        if cotType.hasPrefix("u-d-c-c") { return .target }
        if cotType.hasPrefix("b-") { return .sensor }
        return .unknown
    }

    private func generateCoTXml(event: CoTEvent) -> String {
        let callsign = event.detail.callsign
        let team = event.detail.team ?? "Cyan"

        let dateFormatter = ISO8601DateFormatter()
        let time = dateFormatter.string(from: event.time)
        let start = time
        let stale = dateFormatter.string(from: event.time.addingTimeInterval(3600))

        return """
        <?xml version="1.0" encoding="UTF-8"?>
        <event version="2.0" uid="\(event.uid)" type="\(event.type)" time="\(time)" start="\(start)" stale="\(stale)" how="h-e">
          <point lat="\(event.point.lat)" lon="\(event.point.lon)" hae="\(event.point.hae)" ce="\(event.point.ce)" le="\(event.point.le)"/>
          <detail>
            <contact callsign="\(callsign)"/>
            <__group name="\(team)" role="Team Member"/>
          </detail>
        </event>
        """
    }

    func getConnectedCount() -> Int {
        return servers.filter { $0.status == .connected }.count
    }

    func getTotalCount() -> Int {
        return servers.count
    }

    func clearCache() {
        federatedEvents.removeAll()
        #if DEBUG
        print("🗑️ Event cache cleared")
        #endif
    }
}
